app/post_helpers.go

```diff
@@ -0,0 +1,215 @@
+func linearFilterPosts(postList *model.PostList, earliestAccessibleTime int64) {
+	// filter Posts
+	posts := postList.Posts
+
+	newPostOrder := []string{}
+	for _, postId := range postList.Order {
+		if posts[postId].CreateAt < earliestAccessibleTime {
+			postList.HasInaccessiblePosts = true
+			delete(posts, postId)
+		} else {
+			newPostOrder = append(newPostOrder, postId)
+		}
+	}
+
+	// it can happen that some post list results don't have all posts in the Order field.
+	// for example GetPosts in the CollapsedThreads = false path, parents are not added
+	// to Order
+	for postId := range posts {
+		if posts[postId].CreateAt < earliestAccessibleTime {
+			postList.HasInaccessiblePosts = true
+			delete(posts, postId)
+		}
+	}
+
+	postList.Order = newPostOrder
+}
+
+// filterInaccessiblePosts filters out the posts, past the cloud limit
+func (a *App) filterInaccessiblePosts(postList *model.PostList, options filterPostOptions) *model.AppError {
+	if postList == nil || postList.Posts == nil || len(postList.Posts) == 0 {
+		return nil
+	}
+
+	lastAccessiblePostTime, appErr := a.GetLastAccessiblePostTime()
+	if appErr != nil {
+		return model.NewAppError("filterInaccessiblePosts", "app.last_accessible_post.app_error", nil, appErr.Error(), http.StatusInternalServerError)
+	} else if lastAccessiblePostTime == 0 {
+		// No need to filter, all posts are accessible
+		return nil
+	}
+
+	if len(postList.Posts) == len(postList.Order) && options.assumeSortedCreatedAt {
+		bounds := getTimeSortedPostAccessibleBounds(postList, lastAccessiblePostTime)
+		if bounds.allAccessible() {
+			return nil
+		}
+		if bounds.allInaccessible() {
+			if len(postList.Posts) > 0 {
+				postList.HasInaccessiblePosts = true
+			}
+			postList.Posts = map[string]*model.Post{}
+			postList.Order = []string{}
+			return nil
+		}
+		postList.HasInaccessiblePosts = true
+
+		var otherInaccessibleBound int
+		var otherAccessibleBound int
+		if bounds.accessible > bounds.inaccessible {
+			otherInaccessibleBound = 0
+			otherAccessibleBound = len(postList.Order) - 1
+		} else {
+			otherInaccessibleBound = len(postList.Order) - 1
+			otherAccessibleBound = 0
+		}
+
+		order := postList.Order
+		inaccessibleCount := maxInt(bounds.inaccessible, otherInaccessibleBound) - min(bounds.inaccessible, otherInaccessibleBound)
+		accessibleCount := maxInt(bounds.accessible, otherAccessibleBound) - min(bounds.accessible, otherAccessibleBound)
+		// Linearly cover shorter route
+		if inaccessibleCount < accessibleCount {
+			for i := min(bounds.inaccessible, otherInaccessibleBound); i <= maxInt(bounds.inaccessible, otherInaccessibleBound); i++ {
+				delete(postList.Posts, order[i])
+			}
+		} else {
+			accessiblePosts := make(map[string]*model.Post, accessibleCount+1)
+			for i := min(bounds.accessible, otherAccessibleBound); i <= maxInt(bounds.accessible, otherAccessibleBound); i++ {
+				accessiblePosts[order[i]] = postList.Posts[order[i]]
+			}
+			postList.Posts = accessiblePosts
+		}
```

Comment:
I am failing to understand this piece of logic. From what I understand, we have an ordered slice of posts. And we need to exclude posts that are older than `earliestAccessibleTime`. I am not sure why we need to have this logic about `accessible` and `inAccessible` and doing a binary search. Can you clarify?

Reply:
Since the posts are **sorted**, we don't need to find the `earliestAccessibleTime` _linearly_ and can optimize it via _binary search_.

Reply:
Ok, got that bit. Then I am wondering why do we need a separate struct with `accessible` and `inaccessible` fields for that? Can we just not do a [sort.Search](https://pkg.go.dev/sort#Search) to find the index and then trim everything outside that?

Reply:
Using `sort.Search` would work and I'd be down with that as long as we check for ascending and descending order and use the [right comparator](https://pkg.go.dev/sort#example-Search-DescendingOrder) accordingly. When I recommended the binary search, I didn't know `sort.Search` existed.

Reply:
>  I'd be down with that as long as we check for ascending and descending order

Yes of course, that goes without saying.

> When I recommended the binary search, I didn't know sort.Search existed.

No worries, that's what code reviews are for :)